### PR TITLE
Migrate HorizontalColorbar component to React

### PR DESCRIPTION
**Pull request by fbarl**
Part of https://github.com/tmrowco/electricitymap-contrib/issues/1681#issuecomment-572014604.

#### Tested

Mobile + desktop mode, toggling wind and solar and hovering over map to see markers updated for all 3 colobars.

#### Notes

* There were some parts of the old `HorizontalColobar` component which didn't seem to have been used to I removed them (some ordinal scales code)
* I modified the colobars legend component a little by handing more DOM control from D3 to React
* I simplified the scales in `helpers/scales.js`, namely dropping the complex solar color scale code which wasn't used anywhere and also making max values for colorbars directly encoded in the color scale domains
* I slightly changed the number of ticks displayed with each colorbar, see screenshot below:

image

### CODE REVIEW

File: web/src/components/horizontalcolorbar.js [hunk 1]
```diff
@@ -1,217 +1,83 @@
...
-    const inter = d3.interpolate(e[0], e[1]);
-    const value = inter(i / (10 - 1));
-    return that.originalColorScale(value);
-  });
-  this.scale = d3.scaleLinear()
-    .range(this.colors)
-    .domain(this._domain);
+const HorizontalColorbar = ({
+  colorScale,
+  currentMarker,
```
Comment (corradio) [inline web/src/components/horizontalcolorbar.js]: maybe rename to `currentValue` to make clear it's a float?

File: web/src/components/map.js [hunk 2]
```diff
@@ -112,7 +112,7 @@ export default class Map {
       nonClickableFill: '#D4D9DE',
       oceanColor: '#FAFAFA',
       co2Scale: {
-        steps: [0, 150, 600, 750],
+        steps: [0, 150, 600, 800],
```
Comment (corradio) [inline web/src/components/map.js]: why this change?

File: web/src/helpers/scales.js [hunk 3]
```diff
@@ -1,19 +1,16 @@
...
 import { themes } from './themes';
 
-const d3 = Object.assign(
-  {},
-  require('d3-array'),
-  require('d3-interpolate'),
-  require('d3-scale'),
-  require('d3-scale-chromatic'),
+const uniformSpread = ({ from, to, count }) => (
+  range(count).map(index => (from + (to - from) * index / (count - 1)))
```
Comment (corradio) [inline web/src/helpers/scales.js]: you could consider using d3.range:
https://github.com/d3/d3-array#range

File: web/src/components/map.js (hunk 2 above)
Comment (corradio) [reply web/src/components/map.js]: if I remember correctly we want the 750 to have the same color as all values >750, which is different from having the last color set at 800.

File: web/src/components/horizontalcolorbar.js (hunk 1 above)
Comment (fbarl) [reply web/src/components/horizontalcolorbar.js]: Good point! I fixed it in https://github.com/tmrowco/electricitymap-contrib/pull/2299/commits/8d13810bfe0ca8808c7162cf20c3e441246dc5f2.

FYI I also uncovered a typo there which made me realize that solar layer markers are actually never being displayed - I opened https://github.com/tmrowco/electricitymap-contrib/issues/2305 to address that separately.

File: web/src/components/map.js (hunk 2 above)
Comment (fbarl) [reply web/src/components/map.js]: Ah ok, I see. What I was trying to do there is get rid of the separate constraint constants (in this case `maxCo2`) and make the domain limits be encoded in scales themselves for simplicity.

> we want the 750 to have the same color as all values >750

I made this condition explicitly encoded across all `co2Scale` themes in https://github.com/tmrowco/electricitymap-contrib/pull/2299/commits/79e9082654657f73b50c27098c5cff2a4f995992 - does that make sense?

File: web/src/helpers/scales.js (hunk 3 above)
Comment (fbarl) [reply web/src/helpers/scales.js]: Thanks for the tip, but that didn't make it much nicer due to rounding-off errors and neither D3 nor lodash range including the end value by default.

What I decided to do instead is just be simple and list the domain steps explicitly in https://github.com/tmrowco/electricitymap-contrib/pull/2299/commits/1a2d2acd0c0bc63cad57ff46445808adbcda42b3. I also noticed that _15 m/s_ is very low for an upper cap on wind speed (the marker would often go off the scale), so I doubled it to _30 m/s_, makes sense?

image

**Comment by fbarl** [comment]
@corradio PTAL, I made some changes to the scales.

File: web/src/components/map.js (hunk 2 above)
Comment (corradio) [reply web/src/components/map.js]: perfect

File: web/src/helpers/scales.js (hunk 3 above)
Comment (corradio) [reply web/src/helpers/scales.js]: hmm are you sure it still looks nice visually?

Comment (fbarl) [reply web/src/helpers/scales.js]: It looks different but to me it doesn't look worse - at least values don't go off scale anymore :)

_before_
image

_after_
image

I think the real problem now is with stream lines density, i.e. it would be nice to address https://github.com/tmrowco/electricitymap-contrib/issues/1157.

**Comment by fbarl** [comment]
Thanks for approving @corradio!

I'll merge the PR now and we can keep an eye on the wind scale if any futher adjustments / improvements will be needed.
